### ISSUE TITLE
Handling parallelism in `ocamltest`

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#12631]
(This issue was born as medium-sized internship topic. I am turning it into an issue so that anyone interested can volunteer to work on this.)

`ocamltest` is an internal tool of the compiler distribution that executes test cases in a custom format. It knows how to "list" (find/discover) test files in a given directory, and to "run" a given test file. The glue between these two facets is performed by a shell script (inside a Makefile) that consumes the list of discovered test files and runs `ocamltest` on each of them. A variant uses GNU parallel (a shell tool) to run tests in parallel, reducing the testsuite runtime on multicore machine.

But some of our tests use several cores themselves (some are "burn tests" that try to stress the multicore runtime by saturating the test machine), and running several multicore stress tests in parallel is not a good idea performance-wise. We would need a version of our "parallel" testsuite that knows that certain tests need more than one hardware thread available, and can reduce the amount of parallelism when running multicore stress tests.

(There may be several ways to do this, possibly by using more advanced `parallel` options, or moving to another "standard" tools to run processes in parallel, or implementing a new test runner/driver, possibly directly inside ocamltset.)

**Comment by shindere** [comment]
Many thanks for having open this issue to keep track of the topic.

I am guessing that one bit of the solution will be to properly annotate
tests in terms of the ressources they use. In terms of CPU ressourcs, what
kind of information do  we want to be able to express regarding a test?
Would it be enoough to express the number of cores the test needs, for
instance?

**Comment by xavierleroy** [comment]
> Would it be enoough to express the number of cores the test needs, for instance?

Normally, parallel tests should adapt to the number of cores available, see https://github.com/ocaml/ocaml/blob/trunk/testsuite/HACKING.adoc#dimensioning-the-tests .  So, a first approximation that should work well is to distinguish between
- sequential tests that need only one core and can be run in parallel with other sequential tests, and
- parallel tests that need all the cores and must be run in sequence.

The distinction could be as simple as whether "parallel" occurs in the name of the test (directory or file name).

**Comment by shindere** [comment]
If it's enough to distinguish sequential and parallel tests and if it's
okay to have this distinction based on names, then I think it's more a
matter of test harness (the makefiles) than test driver (ocamltest).

Shall we move all the parallel tests to a distinct subdirectory, or
leave them where they are in `parallel` subdirectories, or leave
themwhere they are with `parallel` in the name of the test (looks less
nice to me)?

Once this has been decided I am assuming that we could have just two targets:
one executing all the tests and one executing only the parallel ones.

All the tests would mean executing the sequential ones first and then
the parallel ones. If ocamltest is enabled, we could even have configure
looking for GNU parallel so that the sequential tests get executed in
parallel by default if GNU parallel is available. How does that sound?

**Comment by xavierleroy** [comment]
A `make parallel` that executes the sequential tests in parallel using GNU `parallel` and the parallel tests in sequence would be a great improvement already.

**Comment by shindere** [comment]
Xavier Leroy (2023/10/25 10:16 -0700):
> A `make parallel` that executes the sequential tests in parallel using
> GNU `parallel` and the parallel tests in sequence would be a great
> improvement already.

We do almost have that -- make parallel executes *all* the tests
inparallel. So the only missing bits are (1) distinguishing these tests
and (2) make them run in sequence.

What do we do? A dedicated `parallel` directory, or make sure those
tests have `parallel` n their names, wherever they are?

**Comment by gasche** [comment]
Some directories have non-parallel and parallel tests for the same feature (for example: weak-ephe-final/, gc-roots/). I find this a natural organization. We could have `parallel/` subdirectory in the test directory, or use `parallel` in the name, but I would avoid having to use a distinct tests/ directory if possible.

(Personally in the long term my preference would go for having this metadata in the TEST block, which would let us gradually move to something finer-grained such as specifying the desired number of domains for the test.)

**Comment by shindere** [comment]
Yes, all what you say makes sense to me, too. At the moment we use the
hierarchy under `testsuite/tests` to group tests by category feature and
the idea to put them else where just because they are parallel does not
look great to me either.

Encoding the information in the TEST block was my initial idea, too,
before reading @xavierleroy's suggestion to encode it somehow in the
path of the test's source file.

Would a boolean `parallel` variable be okay to begin with? ocamltest
would initialize it to false by default and it would be the
responsibility of the parallel tests to set it to true un their root
environment. But then AIUI we will have to first compute lists of tests:
those who are not parallel and can be run in parallel, and those who
are parallel and should be run sequentially.

One other dirty hack we could use would be to have a file named
`parallel` in those directories whose test should  not be run in
parallel and that would make the distinction easy to compute. But it
feels really not clean.

Thoughts?

**Comment by shindere** [comment]
This is still an issue and @MisterDA might be interested in working on it.
